contracts/external/actions/TreasuryAction.sol

```diff
@@ -0,0 +1,116 @@
+import "../../internal/balances/BalanceHandler.sol";
+import "../../internal/balances/TokenHandler.sol";
+import "../../global/StorageLayoutV2.sol";
+import "../../global/Constants.sol";
+import "interfaces/notional/NotionalTreasury.sol";
+import "interfaces/compound/ComptrollerInterface.sol";
+import "interfaces/compound/CErc20Interface.sol";
+import "interfaces/WETH9.sol";
+
+contract TreasuryAction is StorageLayoutV2, NotionalTreasury {
+    using SafeMath for uint256;
+    using SafeInt256 for int256;
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable COMP;
+    Comptroller public immutable COMPTROLLER;
+    address public immutable WETH;
+
+    /// @dev Throws if called by any account other than the owner.
+    modifier onlyOwner() {
+        require(owner == msg.sender, "Ownable: caller is not the owner");
+        _;
+    }
+
+    modifier onlyManagerContract() {
+        require(treasuryManagerContract == msg.sender, "Caller is not the treasury manager");
+        _;
+    }
+
+    constructor(IERC20 _comp, Comptroller _comptroller, address _weth) {
+        COMP = _comp;
+        COMPTROLLER = _comptroller;
+        WETH = _weth;
+        treasuryManagerContract = address(0);
+    }
+
+    function claimCOMP(address[] calldata ctokens) external override onlyManagerContract returns (uint256) {
+        COMPTROLLER.claimComp(address(this), ctokens);
+        uint256 bal = COMP.balanceOf(address(this));
+        COMP.transfer(treasuryManagerContract, bal);
+        return bal;
+    }
+
+    function transferReserveToTreasury(address[] calldata assets)
+        external
+        override
+        onlyManagerContract
+        returns (uint256[] memory)
+    {
+        uint256[] memory amountsTransferred = new uint256[](assets.length);
+
+        for (uint256 i; i < assets.length; i++) {
+            uint16 currencyId = tokenAddressToCurrencyId[assets[i]];
+            require(currencyId != 0, "Token not listed");
+
+            // prettier-ignore
+            (int256 reserve, /* */, /* */, /* */) = BalanceHandler.getBalanceStorage(Constants.RESERVE, currencyId);
+            Token memory token = TokenHandler.getAssetToken(currencyId);
+
+            uint256 totalReserve = reserve.toUint();
+            uint256 totalBalance = IERC20(token.tokenAddress).balanceOf(address(this));
+            uint256 buffer = reserveBuffer[token.tokenAddress];
+
+            // Reserve requirement not defined
+            if (buffer == 0) 
+                continue;
+            
+            uint256 requiredReserve = totalBalance.mul(buffer).div(uint256(Constants.INTERNAL_TOKEN_PRECISION));
+            
+            if (totalReserve > requiredReserve) {
+                uint256 redeemAmount = totalReserve.sub(requiredReserve);
+                Token memory underlying = TokenHandler.getUnderlyingToken(currencyId);
+                amountsTransferred[i] = _redeemCToken(token.tokenAddress, underlying.tokenAddress, redeemAmount);
```

Comment:
Here we should call `TokenHandler.redeem(assetToken, underlyingToken, redeemAmount)`

Reply:
fixed